CLDSRV-780: ServerAccessLogs json schema and functional tests

leif-scality opened the pull request:
- Defined a json schema for the server access logs
- Updated unit and logger output to match the schema
- Implemented functional tests, to check that cloudserver writes logs correctly to the output file

leif-scality committed:
CLDSRV-780: store serverAccessLog.bucketName in api.js

leif-scality committed:
CLDSRV-780: json schema for server access logs

leif-scality committed:
CLDSRV-780: update ServerAccessLogger to follow json schema

leif-scality committed:
CLDSRV-780: add server access logs functional tests

bert-e commented:
# Hello leif-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-780 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.2.5`


Please check the `Fix Version/s` of CLDSRV-780, or the target
branch of this pull request.

leif-scality committed:
CLDSRV-780: update CI for server access logs functional tests

leif-scality committed:
CLDSRV-780: log access logs config at startup

dvasilas commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+            "type": "string"
+        },
+        "startTime": {
+            "description": "Timestamp in seconds.milliseconds, recorded when the server first routes the request. Represents the AWS server access log 'Time' field. String type because to be compatible with clickhouse DateTime64(3) type.",
```
```suggestion
            "description": "Timestamp in seconds.milliseconds, recorded when the server first routes the request. Represents the AWS server access log 'Time' field. String type to be compatible with clickhouse DateTime64(3) type.",
```

leif-scality replied:
done

dvasilas commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+            "const": "0"
+        },
+        "loggingEnabled": {
+            "description": "If true it means that the target bucket has bucket logging configured and enabled.",
```
```suggestion
            "description": "True if the target bucket has bucket logging configured and enabled.",
```

leif-scality replied:
done

dvasilas commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+            "minimum": 0
+        },
+        "requester": {
+            "description": "Represents the AWS server access log 'Requester' field.",
```
```suggestion
            "description": "Represents the AWS server access log 'Requester' field. https://docs.aws.amazon.com/AmazonS3/latest/userguide/LogFormat.html",
```

leif-scality replied:
done

dvasilas commented on tests/functional/aws-node-sdk/test/serverAccessLogs/testServerAccessLogFile.js:
```diff
@@ -0,0 +1,1083 @@
+                const method = async () => {
+                    await s3.createBucket({ Bucket: bucketName }).promise();
+                    await s3.putObject({ Bucket: bucketName, Key: objectKey, Body: 'test data' }).promise();
+                    await s3.copyObject({
```
AWS doc says
> A copy operation involves a GET and a PUT. For that reason, we log two records when performing a copy operation.

Do we have a different behavior for copies?

leif-scality replied:
I am guessing AWS does a GET + PUT in the background to do the copy, I suppose metadata does the copy without recalling the AWS API.

There is only one API call to https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html so only one log entry

The trace 
- 1 PUT
- 1 COPY
- 2 DeleteObject

So the copy is there
```
{"time":1764767027,"hostname":"xps","pid":182,"action":"CreateBucket","accountName":"Bart","accountDisplayName":"Bart","userName":null,"clientPort":41824,"httpMethod":"PUT","bytesDeleted":null,"bytesReceived":0,"bodyLength":0,"contentLength":0,"elapsed_ms":0.27713,"httpURL":"/test-server-access-log-bucket","startTime":"1764767027.212","requester":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","operation":"REST.PUT.BUCKET","requestURI":"PUT /test-server-access-log-bucket HTTP/1.1","errorCode":null,"objectSize":null,"totalTime":"0","turnAroundTime":"0","referer":null,"userAgent":"aws-sdk-nodejs/2.1692.0 linux/v22.14.0 promise","versionID":null,"signatureVersion":"SigV4","cipherSuite":null,"authenticationType":"AuthHeader","hostHeader":"127.0.0.1:8000","tlsVersion":null,"aclRequired":null,"bucketOwner":null,"bucketName":"test-server-access-log-bucket","req_id":"d1608b0bae41b38012da","bytesSent":null,"clientIP":"::ffff:127.0.0.1","httpCode":200,"objectKey":null,"logFormatVersion":"0","loggingEnabled":false,"loggingTargetBucket":null,"loggingTargetPrefix":null,"awsAccessKeyID":"accessKey1","raftSessionID":null}
{"time":1764767027,"hostname":"xps","pid":182,"action":"PutObject","accountName":"Bart","accountDisplayName":"Bart","userName":null,"clientPort":41826,"httpMethod":"PUT","bytesDeleted":null,"bytesReceived":9,"bodyLength":9,"contentLength":9,"elapsed_ms":0.407388,"httpURL":"/test-server-access-log-bucket/test-object-key","startTime":"1764767027.214","requester":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","operation":"REST.PUT.OBJECT","requestURI":"PUT /test-server-access-log-bucket/test-object-key HTTP/1.1","errorCode":null,"objectSize":9,"totalTime":"0","turnAroundTime":null,"referer":null,"userAgent":"aws-sdk-nodejs/2.1692.0 linux/v22.14.0 promise","versionID":null,"signatureVersion":"SigV4","cipherSuite":null,"authenticationType":"AuthHeader","hostHeader":"127.0.0.1:8000","tlsVersion":null,"aclRequired":null,"bucketOwner":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","bucketName":"test-server-access-log-bucket","req_id":"e17dcc57bb40863715d2","bytesSent":null,"clientIP":"::ffff:127.0.0.1","httpCode":200,"objectKey":"test-object-key","logFormatVersion":"0","loggingEnabled":false,"loggingTargetBucket":null,"loggingTargetPrefix":null,"awsAccessKeyID":"accessKey1","raftSessionID":null}
{"time":1764767027,"hostname":"xps","pid":182,"action":"CopyObject","accountName":"Bart","accountDisplayName":"Bart","userName":null,"clientPort":41840,"httpMethod":"PUT","bytesDeleted":null,"bytesReceived":0,"bodyLength":0,"contentLength":0,"elapsed_ms":0.43388,"httpURL":"/test-server-access-log-bucket/test-object-key-copy","startTime":"1764767027.215","requester":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","operation":"REST.PUT.COPY","requestURI":"PUT /test-server-access-log-bucket/test-object-key-copy HTTP/1.1","errorCode":null,"objectSize":null,"totalTime":"0","turnAroundTime":"0","referer":null,"userAgent":"aws-sdk-nodejs/2.1692.0 linux/v22.14.0 promise","versionID":null,"signatureVersion":"SigV4","cipherSuite":null,"authenticationType":"AuthHeader","hostHeader":"127.0.0.1:8000","tlsVersion":null,"aclRequired":null,"bucketOwner":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","bucketName":"test-server-access-log-bucket","req_id":"007c6d149d4cad3261cf","bytesSent":185,"clientIP":"::ffff:127.0.0.1","httpCode":200,"objectKey":"test-object-key-copy","logFormatVersion":"0","loggingEnabled":false,"loggingTargetBucket":null,"loggingTargetPrefix":null,"awsAccessKeyID":"accessKey1","raftSessionID":null}
{"time":1764767027,"hostname":"xps","pid":182,"action":"ListBuckets","accountName":"Bart","accountDisplayName":"Bart","userName":null,"clientPort":41852,"httpMethod":"GET","bytesDeleted":null,"bytesReceived":0,"bodyLength":0,"contentLength":0,"elapsed_ms":0.246037,"httpURL":"/","startTime":"1764767027.217","requester":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","operation":"REST.GET.UNKNOWN","requestURI":"GET / HTTP/1.1","errorCode":null,"objectSize":null,"totalTime":"0","turnAroundTime":"0","referer":null,"userAgent":"aws-sdk-nodejs/2.1692.0 linux/v22.14.0 promise","versionID":null,"signatureVersion":"SigV4","cipherSuite":null,"authenticationType":"AuthHeader","hostHeader":"127.0.0.1:8000","tlsVersion":null,"aclRequired":null,"bucketOwner":null,"bucketName":null,"req_id":"8880cae3184492467b5f","bytesSent":385,"clientIP":"::ffff:127.0.0.1","httpCode":200,"objectKey":null,"logFormatVersion":"0","loggingEnabled":false,"loggingTargetBucket":null,"loggingTargetPrefix":null,"awsAccessKeyID":"accessKey1","raftSessionID":null}
{"time":1764767027,"hostname":"xps","pid":182,"action":"ListMultipartUploads","accountName":"Bart","accountDisplayName":"Bart","userName":null,"clientPort":41860,"httpMethod":"GET","bytesDeleted":null,"bytesReceived":0,"bodyLength":0,"contentLength":0,"elapsed_ms":0.358717,"httpURL":"/test-server-access-log-bucket?uploads","startTime":"1764767027.218","requester":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","operation":"REST.GET.UPLOADS","requestURI":"GET /test-server-access-log-bucket?uploads HTTP/1.1","errorCode":null,"objectSize":null,"totalTime":"0","turnAroundTime":"0","referer":null,"userAgent":"aws-sdk-nodejs/2.1692.0 linux/v22.14.0 promise","versionID":null,"signatureVersion":"SigV4","cipherSuite":null,"authenticationType":"AuthHeader","hostHeader":"127.0.0.1:8000","tlsVersion":null,"aclRequired":null,"bucketOwner":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","bucketName":"test-server-access-log-bucket","req_id":"0c26c04dd745bab9f564","bytesSent":284,"clientIP":"::ffff:127.0.0.1","httpCode":200,"objectKey":null,"logFormatVersion":"0","loggingEnabled":false,"loggingTargetBucket":null,"loggingTargetPrefix":null,"awsAccessKeyID":"accessKey1","raftSessionID":null}
{"time":1764767027,"hostname":"xps","pid":182,"action":"ListObjects","accountName":"Bart","accountDisplayName":"Bart","userName":null,"clientPort":41872,"httpMethod":"GET","bytesDeleted":null,"bytesReceived":0,"bodyLength":0,"contentLength":0,"elapsed_ms":0.327174,"httpURL":"/test-server-access-log-bucket?versions","startTime":"1764767027.219","requester":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","operation":"REST.GET.BUCKET","requestURI":"GET /test-server-access-log-bucket?versions HTTP/1.1","errorCode":null,"objectSize":null,"totalTime":"0","turnAroundTime":"0","referer":null,"userAgent":"aws-sdk-nodejs/2.1692.0 linux/v22.14.0 promise","versionID":null,"signatureVersion":"SigV4","cipherSuite":null,"authenticationType":"AuthHeader","hostHeader":"127.0.0.1:8000","tlsVersion":null,"aclRequired":null,"bucketOwner":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","bucketName":"test-server-access-log-bucket","req_id":"5536b252b849a73e993e","bytesSent":983,"clientIP":"::ffff:127.0.0.1","httpCode":200,"objectKey":null,"logFormatVersion":"0","loggingEnabled":false,"loggingTargetBucket":null,"loggingTargetPrefix":null,"awsAccessKeyID":"accessKey1","raftSessionID":null}
{"time":1764767027,"hostname":"xps","pid":182,"action":"DeleteObject","accountName":"Bart","accountDisplayName":"Bart","userName":null,"clientPort":41880,"httpMethod":"DELETE","bytesDeleted":9,"bytesReceived":0,"bodyLength":0,"contentLength":0,"elapsed_ms":0.380872,"httpURL":"/test-server-access-log-bucket/test-object-key?versionId=null","startTime":"1764767027.220","requester":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","operation":"REST.DELETE.OBJECT","requestURI":"DELETE /test-server-access-log-bucket/test-object-key?versionId=null HTTP/1.1","errorCode":null,"objectSize":null,"totalTime":"0","turnAroundTime":"0","referer":null,"userAgent":"aws-sdk-nodejs/2.1692.0 linux/v22.14.0 promise","versionID":"null","signatureVersion":"SigV4","cipherSuite":null,"authenticationType":"AuthHeader","hostHeader":"127.0.0.1:8000","tlsVersion":null,"aclRequired":null,"bucketOwner":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","bucketName":"test-server-access-log-bucket","req_id":"5d2520863f41a1d58f2d","bytesSent":null,"clientIP":"::ffff:127.0.0.1","httpCode":204,"objectKey":"test-object-key","logFormatVersion":"0","loggingEnabled":false,"loggingTargetBucket":null,"loggingTargetPrefix":null,"awsAccessKeyID":"accessKey1","raftSessionID":null}
{"time":1764767027,"hostname":"xps","pid":182,"action":"DeleteObject","accountName":"Bart","accountDisplayName":"Bart","userName":null,"clientPort":41886,"httpMethod":"DELETE","bytesDeleted":9,"bytesReceived":0,"bodyLength":0,"contentLength":0,"elapsed_ms":0.279938,"httpURL":"/test-server-access-log-bucket/test-object-key-copy?versionId=null","startTime":"1764767027.222","requester":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","operation":"REST.DELETE.OBJECT","requestURI":"DELETE /test-server-access-log-bucket/test-object-key-copy?versionId=null HTTP/1.1","errorCode":null,"objectSize":null,"totalTime":"0","turnAroundTime":"0","referer":null,"userAgent":"aws-sdk-nodejs/2.1692.0 linux/v22.14.0 promise","versionID":"null","signatureVersion":"SigV4","cipherSuite":null,"authenticationType":"AuthHeader","hostHeader":"127.0.0.1:8000","tlsVersion":null,"aclRequired":null,"bucketOwner":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","bucketName":"test-server-access-log-bucket","req_id":"2c7f5939924bada9a4c4","bytesSent":null,"clientIP":"::ffff:127.0.0.1","httpCode":204,"objectKey":"test-object-key-copy","logFormatVersion":"0","loggingEnabled":false,"loggingTargetBucket":null,"loggingTargetPrefix":null,"awsAccessKeyID":"accessKey1","raftSessionID":null}
{"time":1764767027,"hostname":"xps","pid":182,"action":"DeleteBucket","accountName":"Bart","accountDisplayName":"Bart","userName":null,"clientPort":41902,"httpMethod":"DELETE","bytesDeleted":null,"bytesReceived":0,"bodyLength":0,"contentLength":0,"elapsed_ms":0.234398,"httpURL":"/test-server-access-log-bucket","startTime":"1764767027.222","requester":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","operation":"REST.DELETE.BUCKET","requestURI":"DELETE /test-server-access-log-bucket HTTP/1.1","errorCode":null,"objectSize":null,"totalTime":"0","turnAroundTime":"0","referer":null,"userAgent":"aws-sdk-nodejs/2.1692.0 linux/v22.14.0 promise","versionID":null,"signatureVersion":"SigV4","cipherSuite":null,"authenticationType":"AuthHeader","hostHeader":"127.0.0.1:8000","tlsVersion":null,"aclRequired":null,"bucketOwner":"79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be","bucketName":"test-server-access-log-bucket","req_id":"8d56bebe404988d04d4d","bytesSent":null,"clientIP":"::ffff:127.0.0.1","httpCode":204,"objectKey":null,"logFormatVersion":"0","loggingEnabled":false,"loggingTargetBucket":null,"loggingTargetPrefix":null,"awsAccessKeyID":"accessKey1","raftSessionID":null}
```

dvasilas replied:
Makes sense.

dvasilas commented on tests/functional/aws-node-sdk/test/serverAccessLogs/testServerAccessLogFile.js:
```diff
@@ -0,0 +1,1083 @@
+                    const result = tv4.validateResult(logEntries[logEntryIdx], schema);
+                    assert.strictEqual(result.valid, true,
+                        `Log entry should match schema: ${JSON.stringify(result.error)}`);
+                    assert.strictEqual(logEntries[logEntryIdx].operation, operation.expectedOperations[operationIdx],
```
Do you think we can add checks for more fields (maybe not this PR) ?
I am thinking at least `bytesReceived`, `bytesDeleted`, `objectSize`, `versionID`.

There are some checks that you can probably add easily in this PR: `bucketName`, `bucketOwner` (?), `objectKey`, `loggingEnabled`, `loggingTargetBucket`, `loggingTargetPrefix`, `awsAccessKeyID`, `httpMethod`, `httpCode`.

My point is: let's try to have as much test coverage as possible for the log record.

leif-scality replied:
I now check more properties.
- Time properties are not tested because we can't control them, the type is checked by the schema
- Bytes*/contentLength are not tested because I think they may change so better do it after. The schema checks their types
- VersionID we don't know what version value is going to be used 
- TLS is not tested because we don't have an HTTPS setup
```js
        // List of all the log properties.
        // Commented properties are set by the test or not tested.
        // UNKNOWN: The property is not accessible by the test or is non determistic.
        // DYNAMIC: The propery is set depending on the operation.
        // STATIC: Shared among all tests.
        // TODO: The property need to be tested.
        const commonProperties = {
            // 'time': '', // UNKNOWN
            // 'hostname': '', // UNKNOWN
            // 'pid': '', // UNKNOWN
            'action': 'REQUIRED', // DYNAMIC
            'accountName': 'Bart', // STATIC
            'accountDisplayName': 'Bart', // STATIC
            'userName': null, // TODO: Add test with IAM user to get a non null userName.
            // 'clientPort': '', // UNKNOWN
            'httpMethod': 'REQUIRED', // DYNAMIC
            // 'bytesDeleted': '', // TODO
            // 'bytesReceived': '', // TODO
            // 'bodyLength': '', // TODO
            // 'contentLength': '', // TODO
            // 'elapsed_ms': '', // UNKNOWN
            // 'httpURL': '', // TODO
            // 'startTime': '', // UNKNOWN
            'requester': '79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be', // STATIC
            'operation': 'REQUIRED', // DYNAMIC
            // 'requestURI': '', // TODO
            'errorCode': null,  // DYNAMIC
            // 'objectSize': '', // TODO
            // 'totalTime': '', // UNKNOWN
            // 'turnAroundTime': '', // UNKNOWN
            'referer': null, // TODO: Add test that sets the referer.
            // 'userAgent': // UNKNOWN
            // 'versionID': '', // UNKNOWN
            'signatureVersion': 'SigV4', // STATIC
            'cipherSuite': null, // TODO: Add https tests.
            'authenticationType': 'AuthHeader', // STATIC
            // 'hostHeader': '', // UNKNOWN
            'tlsVersion': null, // TODO: Add https tests.
            'aclRequired': null, // TODO: Add https tests.
            'bucketOwner': '79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be', // DYNAMIC
            bucketName, // DYNAMIC
            // 'req_id': '', // UNKNOWN
            // 'bytesSent': '', // TODO
            // 'clientIP': '', // UNKNOWN
            'httpCode': 200, // DYNAMIC
            'objectKey': null, // DYNAMIC
            'logFormatVersion': '0', // STATIC
            'loggingEnabled': false, // DYNAMIC
            'loggingTargetBucket': null, // DYNAMIC
            'loggingTargetPrefix': null, // DYNAMIC
            'awsAccessKeyID': 'accessKey1', // STATIC
            'raftSessionID': null, // UNKNOWN
        };
```

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-780 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.2.6`


Please check the `Fix Version/s` of CLDSRV-780, or the target
branch of this pull request.

dvasilas commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+        "startTime": {
+            "description": "Timestamp in seconds.milliseconds, recorded when the server first routes the request. Represents the AWS server access log 'Time' field. String type to be compatible with clickhouse DateTime64(3) type.",
+            "type": "string",
+            "minimum": 0
```
Minimum is probably not needed here.

leif-scality replied:
The startTime cannot be negative, so it is just an extra sanity check

dvasilas replied:
Yes, I mean that it may not apply because statTime is a string. Not sure.

leif-scality replied:
Forgot it was a string, you are right

leif-scality replied:
fixed

dvasilas reviewed (approved):
Don't you need an Arsenal bump?

tcarmet commented on .github/docker/docker-compose.yaml:
```diff
@@ -8,6 +8,7 @@ services:
       - ${HOME}/.aws/credentials:/root/.aws/credentials
       - /tmp/artifacts/${JOB_NAME}:/artifacts
       - /tmp/coverage/${JOB_NAME}:/coverage/test
+      - /logs:/logs
```
I understand why you are doing this, but it's a bit annoying when we want to execute those locally to either remind ourselves to create this file, or don't forget to add the steps you added in the CI to make sure we create the `/logs` directory.

Is there a way to reduce a bit those? either use `/tmp/logs` on the host or a directory in the repo that you keep with `.gitkeep` but ignore everything inside with `.gitignore` configuration. Just so that's it's a bit less problematic for local execution.

leif-scality replied:
The tests read the `config.json` file so you can modify `serverAccessLogs.outputFile` to read from where ever you want.
I wrote the CI code for it to work with the default config. But if you change the config and both the tests and the server are in sync it should work

tcarmet replied:
So having something like `/tmp/logs:/logs` would break it, because both cloudserver and the tests need to read the logs from the same file path?

leif-scality replied:
In the CI if you don't change the test config.json to read from `/tmp/logs` and create the `/tmp/logs` folder, yes it will break.

In the CI the clouserver container and the tests don't read the same config file, so cloudserver would write to its `/logs` in its container, and the changes would be visible in `/tmp/logs` in the host filesystem. For this to work you would have to update the test config.json for it to read from `/tmp/logs` and create the `/tmp/logs` folder before the test.

The only way for the test to work locally and in the CI is to change the config.json to `/tmp/logs` but then you need to change the value in Federation back to `/logs` and you also have to create the folder in all cases because cloudserver  only creates the file not the folder.

tcarmet replied:
okkk got it, thank you for the clarification! ❤️

leif-scality committed:
CLDSRV-780: bump arsenal to 8.2.42

jonathan-gramain commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+{
+    "$id": "https://scality.com/cloudserver/server_access_log.schema.json",
```
I would rather see a github.com link pointing to the actual file in a given branch of the Cloudserver repository here

leif-scality replied:
replaced with "https://github.com/scality/cloudserver/tree/development/9.2/schema/lib/server_access_log.schema.json"

jonathan-gramain commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+    "type": "object",
+    "properties": {
+        "time": {
+            "description": "Epoch timestamp in seconds, recorded when the log record is created.",
```
nitpick: some descriptions end with a period, some not, they should be consistent. I would think no period is better if they are not complete sentences.

leif-scality replied:
Added periods to all the lines

jonathan-gramain commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+            "type": "string"
+        },
+        "accountName": {
+            "description": "Requester account display name.",
```
```suggestion
            "description": "Requester account name",
```

leif-scality replied:
Changed to the correct description

jonathan-gramain commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+            "enum": ["GET", "HEAD", "OPTIONS", "TRACE", "PUT", "DELETE", "POST", "PATCH", "CONNECT"]
+        },
+        "bytesDeleted": {
+            "description": "For DeleteObject bytesDeleted is the size in bytes of the deleted object, for DeleteObjects it is the sum of all the deleted objects.",
```
suggestion for a bit less verbosity and more accuracy:

```suggestion
            "description": "For DeleteObject: size in bytes of the deleted object, for DeleteObjects: byte size sum of all the deleted objects",
```

leif-scality replied:
done

jonathan-gramain commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+            "minimum": 0
+        },
+        "bytesReceived": {
+            "description": "Size of the object in bytes of PutObject and UploadPart",
```
```suggestion
            "description": "For PutObject and UploadPart: size of the object in bytes",
```

leif-scality replied:
done

jonathan-gramain commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+            "type": "string"
+        },
+        "startTime": {
+            "description": "Timestamp in seconds.milliseconds, recorded when the server first routes the request. Represents the AWS server access log 'Time' field. String type to be compatible with clickhouse DateTime64(3) type.",
```
```suggestion
            "description": "Timestamp formatted as: 'seconds.milliseconds', recorded when the server first routes the request. Represents the AWS server access log 'Time' field. String type compatible with Clickhouse DateTime64(3) type.",
```

leif-scality replied:
done

jonathan-gramain commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+            "minimum": 0
+        },
+        "requester": {
+            "description": "Represents the AWS server access log 'Requester' field. https://docs.aws.amazon.com/AmazonS3/latest/userguide/LogFormat.html",
```
I think you could get rid of "Represents the" for this and all following values, should be self explanatory without it.

leif-scality replied:
done

jonathan-gramain commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+            "type": "string"
+        },
+        "tlsVersion": {
+            "description": "Represents the AWS server access log 'Tls Version' field.",
```
```suggestion
            "description": "Represents the AWS server access log 'TLS Version' field.",
```

leif-scality replied:
done

jonathan-gramain commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+            "type": "string"
+        },
+        "req_id": {
+            "description": "Represents the AWS server access log 'Request ID' field. Matches the req_id of the stdout logs.",
```
I think "stdout" is an implementation detail (and it ends up in log files on S3C transparently) so I'd rather just say "matches the req_id field in logs"

leif-scality replied:
done

jonathan-gramain commented on lib/utilities/serverAccessLogger.js:
```diff
@@ -182,74 +182,75 @@ function getRemoteIPFromRequest(request) {
-        'listParts': 'OBJECT',
+        'completeMultipartUpload': 'UPLOAD',
+        'initiateMultipartUpload': 'UPLOAD',
+        'listMultipartUploads': 'UPLOADS',
```
Checking if it's not a typo and you mean plural here, since others are all singular.

leif-scality replied:
It is upload in this case, but I fixed other issues in the map

https://github.com/awslabs/glue-extensions-for-iceberg/blob/52bdb2908216a85859fd76a45981d0326d016a2f/spark/src/main/scala/software/amazon/glue/s3a/audit/S3LogVerbs.java#L36

jonathan-gramain commented on tests/functional/aws-node-sdk/test/serverAccessLogs/testServerAccessLogFile.js:
Nitpick: generally speaking, it should be doable to merge tests for groups of API functions that put, get and delete something, in a single test that would put, then get, then delete the API object. Just to reduce the number of tests since I believe the test coverage would be identical.

Not blocking for me, just a suggestion.

jonathan-gramain commented on tests/functional/aws-node-sdk/test/serverAccessLogs/testServerAccessLogFile.js:
```diff
@@ -0,0 +1,2562 @@
+                    assert.strictEqual(result.valid, true,
+                        `Log entry should match schema: ${JSON.stringify(result.error)}`);
+
+                    const properties = operation.expected[operationIdx];
+                    for (const [key, val] of Object.entries(properties)) {
+                        assert.strictEqual(logEntries[logEntryIdx][key], val,
+                            `Invalid value for ${key}, action ${properties.action}`);
+                    }
```
Could this be replaced by:

```
assert.deepStrictEqual(logEntries[logEntryIdx], operation.expected[operationIdx])
```

A failed assertion would show a diff, so I think it would give the same amount of debug info, if not more.

leif-scality replied:
No, because some fields we don't control like timestamps.

Output with the suggested change 
```
+ actual - expected

  {
    accountDisplayName: 'Bart',
    accountName: 'Bart',
    aclRequired: null,
    action: 'DeleteBucket',
    authenticationType: 'AuthHeader',
    awsAccessKeyID: 'accessKey1',
+   bodyLength: 0,
    bucketName: 'xxx',
    bucketOwner: null,
+   bytesDeleted: null,
+   bytesReceived: 0,
+   bytesSent: 197,
    cipherSuite: null,
+   clientIP: '::ffff:127.0.0.1',
+   clientPort: 52912,
+   contentLength: 0,
+   elapsed_ms: 0.32016,
    errorCode: 'NoSuchBucket',
+   hostHeader: '127.0.0.1:8000',
+   hostname: 'xps',
    httpCode: 404,
    httpMethod: 'DELETE',
+   httpURL: '/xxx',
    logFormatVersion: '0',
    loggingEnabled: false,
    loggingTargetBucket: null,
    loggingTargetPrefix: null,
    objectKey: null,
+   objectSize: null,
    operation: 'REST.DELETE.BUCKET',
+   pid: 10496,
    raftSessionID: null,
    referer: null,
+   req_id: '5061ab0fdc4b9cf4db78',
+   requestURI: 'DELETE /xxx HTTP/1.1',
    requester: '79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be',
    signatureVersion: 'SigV4',
+   startTime: '1764951661.826',
+   time: 1764951661,
    tlsVersion: null,
+   totalTime: '0',
+   turnAroundTime: '0',
+   userAgent: 'aws-sdk-nodejs/2.1692.0 linux/v22.14.0 promise',
    userName: null,
+   versionID: null
  }

      + expected - actual

         "aclRequired": [null]
         "action": "DeleteBucket"
         "authenticationType": "AuthHeader"
         "awsAccessKeyID": "accessKey1"
      -  "bodyLength": 0
         "bucketName": "xxx"
         "bucketOwner": [null]
      -  "bytesDeleted": [null]
      -  "bytesReceived": 0
      -  "bytesSent": 197
         "cipherSuite": [null]
      -  "clientIP": "::ffff:127.0.0.1"
      -  "clientPort": 52912
      -  "contentLength": 0
      -  "elapsed_ms": 0.32016
         "errorCode": "NoSuchBucket"
      -  "hostHeader": "127.0.0.1:8000"
      -  "hostname": "xps"
         "httpCode": 404
         "httpMethod": "DELETE"
      -  "httpURL": "/xxx"
         "logFormatVersion": "0"
         "loggingEnabled": false
         "loggingTargetBucket": [null]
         "loggingTargetPrefix": [null]
         "objectKey": [null]
      -  "objectSize": [null]
         "operation": "REST.DELETE.BUCKET"
      -  "pid": 10496
         "raftSessionID": [null]
         "referer": [null]
      -  "req_id": "5061ab0fdc4b9cf4db78"
      -  "requestURI": "DELETE /xxx HTTP/1.1"
         "requester": "79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be"
         "signatureVersion": "SigV4"
      -  "startTime": "1764951661.826"
      -  "time": 1764951661
         "tlsVersion": [null]
      -  "totalTime": "0"
      -  "turnAroundTime": "0"
      -  "userAgent": "aws-sdk-nodejs/2.1692.0 linux/v22.14.0 promise"
         "userName": [null]
      -  "versionID": [null]
       }
      
```

jonathan-gramain commented on lib/server.js:
```diff
@@ -396,7 +396,8 @@ class S3Server {
 
-           try {
+            try {
+                logger.info('ServerAccessLogger config', { 'config': _config.serverAccessLogs });
```
```suggestion
                logger.info('ServerAccessLogger config', { config: _config.serverAccessLogs });
```

leif-scality replied:
fixed

nicolas2bert commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,248 @@
+            "description": "Represents the AWS server access log 'ACL Required' field.",
+            "type": ["null"]
+        },
+        "bucketOwner": {
```
`bucketOwner` and `bucketName` should never be `null`.
This is what is breaking integration tests.

For some operations like CreateBucket, deleteObject**s**, uploadPart, the bucket

here are the server access logs for:
- deleteobjects:
```    
{"time":1764928092,"hostname":"nicolas2be-952-store-4","pid":81,"action":"DeleteObjects","accountName":"jul","accountDisplayName":"jul","userName":null,"clientPort":51762,"httpMethod":"POST","bytesDeleted":null,"bytesReceived":129,"bodyLength":129,"contentLength":129,"elapsed_ms":183.005601,"httpURL":"/bucket0?delete","startTime":"1764928092.696","requester":"b4e50bcfa5a588f74c82becb4476a0d7442892ba27d05baa847f0528b1076f72","operation":"REST.POST.OBJECT","requestURI":"POST /bucket0?delete HTTP/1.1","errorCode":null,"objectSize":null,"totalTime":"182","turnAroundTime":"100","referer":null,"userAgent":"aws-cli/1.38.10 md/Botocore#1.37.10 ua/2.1 os/macos#24.5.0 md/arch#arm64 lang/python#3.13.2 md/pyimpl#CPython m/N cfg/retry-mode#legacy botocore/1.37.10","versionID":null,"signatureVersion":"SigV4","cipherSuite":null,"authenticationType":"AuthHeader","hostHeader":"51.21.225.168:8080","tlsVersion":null,"aclRequired":null,"bucketOwner":null,"bucketName":null,"req_id":"151a943d0741b0e6144c","bytesSent":183,"clientIP":"10.160.120.218","httpCode":200,"objectKey":null,"logFormatVersion":"0","loggingEnabled":false,"loggingTargetBucket":null,"loggingTargetPrefix":null,"awsAccessKeyID":"GRRX6ENBPIXH0XXUPZ1P","raftSessionID":null}
```

- CreateBucket:

```
{"time":1764841467,"hostname":"nicolas2be-952-store-1","pid":109,"action":"CreateBucket","accountName":"jul","accountDisplayName":"jul","userName":null,"clientPort":31794,"httpMethod":"PUT","bytesDeleted":null,"bytesReceived":0,"bodyLength":0,"contentLength":0,"elapsed_ms":268.683911,"httpURL":"/bucket0","startTime":"1764841467.173","requester":"b4e50bcfa5a588f74c82becb4476a0d7442892ba27d05baa847f0528b1076f72","operation":"REST.PUT.BUCKET","requestURI":"PUT /bucket0 HTTP/1.1","errorCode":null,"objectSize":null,"totalTime":"268","turnAroundTime":"224","referer":null,"userAgent":"aws-cli/1.38.10 md/Botocore#1.37.10 ua/2.1 os/macos#24.5.0 md/arch#arm64 lang/python#3.13.2 md/pyimpl#CPython m/N cfg/retry-mode#legacy botocore/1.37.10","versionID":null,"signatureVersion":"SigV4","cipherSuite":null,"authenticationType":"AuthHeader","hostHeader":"51.21.225.168:8080","tlsVersion":null,"aclRequired":null,"bucketOwner":null,"bucketName":null,"req_id":"cbfa47daf640b6b6ec00","bytesSent":null,"clientIP":"10.160.120.218","httpCode":200,"objectKey":null,"logFormatVersion":"0","loggingEnabled":false,"loggingTargetBucket":null,"loggingTargetPrefix":null,"awsAccessKeyID":"GRRX6ENBPIXH0XXUPZ1P","raftSessionID":null}
```

leif-scality replied:
Fixed by b361882ec996f2b5e1be60e79974cebfe916e674

leif-scality replied:
> bucketOwner and bucketName should never be null.

For `ListBuckets` they are going to be `null`

leif-scality committed:
CLDSRV-780: add bucket owner to server access log in multiObjectDelete

dvasilas reviewed (commented):
We should also include tests about error cases (`AccessDenied`, `NoSuchBucket`, ..).

leif-scality commented:
> We should also include tests about error cases (`AccessDenied`, `NoSuchBucket`, ..).

I added one test with `NoSuchBucket` to check the `errorCode` is set

leif-scality commented:
ping

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-780 contains:


* `9.2.5`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.2.6`


Please check the `Fix Version/s` of CLDSRV-780, or the target
branch of this pull request.

leif-scality committed:
CLDSRV-780: Bump version to 9.2.6

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

leif-scality commented:
ping

leif-scality commented:
> > We should also include tests about error cases (`AccessDenied`, `NoSuchBucket`, ..).
> 
> I added one test with `NoSuchBucket` to check the `errorCode` is set

Also added error test for PutObject and GetObject

anurag4DSB commented on lib/api/api.js:
```diff
@@ -182,6 +182,7 @@ const api = {
             });
         }
         if (request.serverAccessLog) {
+            request.serverAccessLog.bucketName = request.bucketName;
```
Curious about the reason for this move.
Why did we change the location where we add the bucket name to server access logs?

leif-scality replied:
Calls like CreateBucket don't call the helpers so bucketName was always null.

https://github.com/scality/cloudserver/pull/6020#discussion_r2597646221

anurag4DSB commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,247 @@
+        },
+        "turnAroundTime": {
+            "description": "AWS server access log 'Turn Around Time' field.",
+            "type": ["string", "null"]
+        },
+        "referer": {
+            "description": "AWS server access log 'Referer' field.",
+            "type": ["string", "null"]
+        },
+        "userAgent": {
+            "description": "AWS server access log 'User Agent' field.",
+            "type": ["string", "null"]
+        },
+        "versionID": {
+            "description": "AWS server access log 'Version ID' field.",
+            "type": ["string", "null"]
+        },
+        "signatureVersion": {
+            "description": "AWS server access log 'Signature Version' field.",
+            "type": ["string", "null"]
+        },
+        "cipherSuite": {
+            "description": "AWS server access log 'Cipher Suite' field.",
+            "type": ["string", "null"]
+        },
+        "authenticationType": {
+            "description": "AWS server access log 'Authentication Type' field.",
+            "type": ["string", "null"]
+        },
+        "hostHeader": {
+            "description": "AWS server access log 'Host Header' field.",
+            "type": "string"
+        },
+        "tlsVersion": {
+            "description": "AWS server access log 'TLS Version' field.",
+            "type": ["string", "null"]
+        },
+        "aclRequired": {
+            "description": "AWS server access log 'ACL Required' field.",
+            "type": ["null"]
```
For all the "AWS server access logs 'XXX' field." descriptions, have we considered using actual descriptions as AWS docs change without version control? Or is it intentional that we want engineers to refer to https://docs.aws.amazon.com/AmazonS3/latest/userguide/LogFormat.html for what the fields actually represent?
It's worth adding the actual description, and we can add a code comment that these are AWS server access log fields.
Example for `aclRequired`, `AWS server access log 'aclRequired' field indicating whether ACL authorization was required.` would be more helpful than `"AWS server access log 'ACL Required' field."`

Also if we want to be AWS compatible, the proper description will also help us check if we are actually compatible. For example for `aclRequired` allows null, but the actual AWS schema should accept "Yes" or "-".

Example correction:
```
"aclRequired": {
    "description": "AWS server access log 'aclRequired' field indicating whether ACL authorization was required.",
    "type": "string",
    "enum": ["Yes", "-"]
}
```

if we want to allow nulls in our own internal representation, we could explicitly combine them with something like this

```
"aclRequired": {
    "description": "AWS server access log 'aclRequired' field indicating whether ACL authorization was required.",
    "type": ["string", "null"],
    "enum": ["Yes", "-"]
}
```

anurag4DSB replied:
I will let you check and change for others as well.

leif-scality replied:
inlined descriptions

anurag4DSB commented on tests/functional/aws-node-sdk/test/serverAccessLogs/testServerAccessLogFile.js:
```diff
@@ -0,0 +1,2645 @@
+            try {
+                return lines.map(line => JSON.parse(line));
+            } catch (err) {
+                // FIXME: readFileSync may read partial lines making JSON.parse fail, so we need to retry.
```
The FIXME comments acknowledge that readFileSync may read partial lines, causing JSON.parse to fail. This is a real race condition that will cause flaky tests.

We could use a more robust approach with file position tracking

leif-scality replied:
We prevent flaky test by retrying, cloudserver should not take more than 5 seconds to write the logs 
I already create a ticket to track this  CLDSRV-800

anurag4DSB replied:
Could we add the ticket number to the comment please?
(for people just looking at the code)

leif-scality replied:
done

anurag4DSB commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,247 @@
+            "type": "string",
+            "enum": ["GET", "HEAD", "OPTIONS", "TRACE", "PUT", "DELETE", "POST", "PATCH", "CONNECT"]
+        },
+        "bytesDeleted": {
```
`bytesDeleted`, `bytesReceived`, `bodyLength`, `contentLength` all have type: `["integer", "null"]` with minimum: 0, wouldn't null values fail the minimum constraint?
We could do something like this":
```
 {
      "bytesDeleted": {
          "description": "For DeleteObject: size in bytes...",
          "oneOf": [
              {"type": "integer", "minimum": 0},
              {"type": "null"}
          ]
      }
  }
```

Or better, make these fields optional and omit them when null.

leif-scality replied:
The minimum is only applied when the value is an integer

anurag4DSB commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,247 @@
+{
```
For future iterations, we could also add a pattern for these fields.

leif-scality replied:
?

anurag4DSB commented on tests/functional/aws-node-sdk/test/serverAccessLogs/testServerAccessLogFile.js:
```diff
@@ -0,0 +1,2645 @@
+const assert = require('assert');
```
We should add tests for fields `versionID` populated, `cipherSuite` and `tlsVersion`(if possible)

For my edification,  what happens when we log-rotate mid-request?

leif-scality replied:
> We should add tests for fields versionID populated, cipherSuite and tlsVersion(if possible)

cipherSuite and tlsVersion require the client connect using https, I don't know if we  currently have the setup to do it. 

> For my edification, what happens when we log-rotate mid-request?

cloudserver keeps writing to the old file until it detects the new file, fluent bit also keeps reading the old file so we don't lose logs. By default cloudserver checks for file changes every 10secs and fluentbit keeps reading for 30seconds, so we should switch to the new file before fluentbit stops reading

anurag4DSB commented on lib/utilities/serverAccessLogger.js:
```diff
@@ -182,74 +182,78 @@ function getRemoteIPFromRequest(request) {
+        'bucketPutEncryption': 'ENCRYPTION',
         'bucketPutLogging': 'LOGGING_STATUS',
         'bucketGetLogging': 'LOGGING_STATUS',
         // 'corsPreflight': '',
```
WHy is this commented out?

anurag4DSB replied:
Could we add a fallback or/and a JIra ticket here?
`methodToResType[req.apiMethod] || 'UNKNOWN'`

leif-scality replied:
corsPreflight is not an S3 operation so it will output `UNKNOWN`

anurag4DSB commented on lib/utilities/serverAccessLogger.js:
```diff
@@ -182,74 +182,78 @@ function getRemoteIPFromRequest(request) {
+        'websiteHead': 'WEBSITE',
     });
 
     return `REST.${req.method}.${methodToResType[req.apiMethod] ? methodToResType[req.apiMethod] : 'UNKNOWN'}`;
```
Out of scope for this PR, but as I faced it as well, if req.apiMethod is not in the map, it logs REST.GET.UNKNOWN which is valid per schema, but doesn't help debugging. We should log a warning.

something like this
```
.
.
.
      const resourceType = methodToResType[req.apiMethod];
      if (!resourceType) {
          logger.warn('Unknown apiMethod for server access log', {
              apiMethod: req.apiMethod,
              method: req.method,
              url: req.url,
          });
          return `REST.${req.method}.UNKNOWN`;
      }

      return `REST.${req.method}.${resourceType}`;
  }
```

BourgoisMickael replied:
I wonder if we should even use `process.emitWarning` to produce the warning directly into stderr to have more chance to keep it for long and notice it rather than having it disappear in regular log rotation.

As we should notice and fix this find of issue fast

leif-scality replied:
added `process.emitWarning`

anurag4DSB commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,247 @@
+{
+    "$id": "https://github.com/scality/cloudserver/tree/development/9.2/schema/lib/server_access_log.schema.json",
```
Another nit pick: The schema doesn't document which fields are populated for which operations. For example:
  - objectKey is null for bucket operations
  - bytesDeleted is only set for DELETE operations
  - versionID is only set for versioned bucket requests

we could add a top-level examples array showing complete log entries for different operation types, or add conditional validation using if/then/else:
```
  {
      "$id": "...",
      "$schema": "...",
      "title": "Server Access Log",
      "description": "A server access log",
      "type": "object",
      "properties": { /* ... */ },
      "required": [ /* ... */ ],
      "examples": [
          {
              "description": "Example PutObject log",
              "value": { /* complete log entry */ }
          },
          {
              "description": "Example GetObject log",
              "value": { /* complete log entry */ }
          }
      ]
  }
```

anurag4DSB commented on tests/functional/aws-node-sdk/test/serverAccessLogs/testServerAccessLogFile.js:
```diff
@@ -0,0 +1,2645 @@
+    }
+}
+
+async function waitForLogs(filePath, expectedLines, maxRetries = 50, delayMs = 100) {
```
There are magic numbers here, lets explicitly extract these at module level, 
```
  const TEST_CONFIG = {
      MAX_LOG_WAIT_RETRIES: 50,
      LOG_POLL_DELAY_MS: 100,
      LOG_READ_TIMEOUT_MS: 5000,
  };
```

leif-scality replied:
added constant

anurag4DSB commented on tests/functional/aws-node-sdk/test/serverAccessLogs/testServerAccessLogFile.js:
```diff
@@ -0,0 +1,2645 @@
+                    method,
+                    methodName: 'bucketDeleteCors',
+                    expected: [
+                        {
+                            ...commonProperties,
+                            operation: 'REST.PUT.BUCKET',
+                            bucketOwner: null,
+                            action: 'CreateBucket',
+                            httpMethod: 'PUT',
+                        },
+                        {
+                            ...commonProperties,
+                            operation: 'REST.PUT.CORS',
+                            action: 'PutBucketCors',
+                            httpMethod: 'PUT',
+                        },
+                        {
+                            ...commonProperties,
+                            operation: 'REST.DELETE.CORS',
+                            action: 'DeleteBucketCors',
+                            httpCode: 204,
+                            httpMethod: 'DELETE',
```
NIT: There is a lot of code duplication, worth extracting into a helper function.
Something like this
```
function expectedCreateBucket(overrides = {}) {
      return {
          ...commonProperties,
          operation: 'REST.PUT.BUCKET',
          bucketOwner: null,
          action: 'CreateBucket',
          httpMethod: 'PUT',
          ...overrides,
      };
  }

  // Usage:
  expected: [
      expectedCreateBucket(),
      expectedPutObject({ objectKey }),
      expectedDeleteObject({ objectKey, httpCode: 204 }),
  ]
```

leif-scality replied:
Yeah looking back we can do better, the log format should change so I can add the improvements when I update the tests in the next iteration

anurag4DSB commented on tests/functional/aws-node-sdk/test/serverAccessLogs/testServerAccessLogFile.js:
```diff
@@ -0,0 +1,2645 @@
+        });
+
+        for (const operation of operations) {
+            it(`check ${operation.methodName} log entry`, async () => {
```
Worth being more descriptive here

```suggestion
            it(`should log correct ${operation.methodName} operation with all required fields`, async () => {
```

leif-scality replied:
done

anurag4DSB commented on schema/server_access_log.schema.json:
```diff
@@ -0,0 +1,247 @@
+        "awsAccessKeyID",
+        "raftSessionID"
+    ]
+}
```
The schema doesn't specify "additionalProperties": false, allowing undocumented fields to be added.
This ensures strict validation and prevents accidental field additions.
```
  {
      "properties": { /* ... */ },
      "required": [ /* ... */ ],
      "additionalProperties": false
  }
```

leif-scality replied:
done

anurag4DSB commented on tests/functional/aws-node-sdk/test/serverAccessLogs/testServerAccessLogFile.js:
```diff
@@ -0,0 +1,2645 @@
+        bu = bucketUtil;
+        const s3 = bucketUtil.s3;
+        const logFilePath = config.serverAccessLogs.outputFile;
+        const bucketName = 'test-server-access-log-bucket';
```
If tests run in parallel or are interrupted, could this cause conflicts?

leif-scality replied:
The tests expect the server-access.log file to be empty, if tests are run in parallel this is no longer true so they will break.

anurag4DSB commented on lib/utilities/serverAccessLogger.js:
```diff
@@ -387,58 +391,58 @@ function logServerAccess(req, res) {
+        httpMethod: req.method === undefined ? null : req.method,
+        bytesDeleted: params.analyticsBytesDeleted === undefined ? null : params.analyticsBytesDeleted,
+        bytesReceived: req.parsedContentLength === undefined ? null : req.parsedContentLength,
+        bodyLength: req.headers['content-length'] === undefined ? null : parseInt(req.headers['content-length'], 10),
```
Very nitpicking here: There are some hard-to-read and sometimes inconsistent complex null coalescing here.
Could we standardize them?
example: 

```
  accountDisplayName: authInfo?.getAccountDisplayName() ?? null,
  clientPort: req.socket?.remotePort ?? null,
  bytesDeleted: params.analyticsBytesDeleted ?? null,
```

anurag4DSB reviewed (dismissed):
As this is a new standard which will be used by future updates to S3 analytics, we should also write a document to guide engineers on whats the right way to add new fields to server access logs. This will be helpful for everyone to adhere to the standards thought by the creators of server access logs.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers






The following reviewers are expecting changes from the author, or must review again:

* @anurag4dsb

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

anurag4DSB commented:
Removing blocker, as per discussion in daily, might be addressed later in other tickets.

BourgoisMickael commented on lib/utilities/serverAccessLogger.js:
```diff
@@ -301,7 +305,7 @@ function getObjectSize(request, response) {
         }
 
-        return len;
+        return Number(len);
```
It's not very nice to encapsulate a primitive into its object constructor instance.

Besides the constructor might convert `''` to `0` instead of `NaN`. Is it the expected behavior to default to 0 for empty string ?

leif-scality replied:
Changed to 
```
    if (request && response && objectSizeGetMethods[request.apiMethod]) {
        const len = Number(response.getHeader('Content-Length'));
        if (len === undefined || len === null || len === '') {
            return null;
        }

        const numLen = Number(len);
        if (isNaN(numLen)) {
            return null;
        }

        return numLen;
    }

```

BourgoisMickael replied:
That doesn't change the use of `Number`, this double the usage. And `Number` cannot return null or undefined. It returns a number

BourgoisMickael commented on lib/utilities/serverAccessLogger.js:
```diff
@@ -301,7 +305,7 @@ function getObjectSize(request, response) {
+        return Number(len);
     }
 
     if (request && objectSizePutMethods[request.apiMethod]) {
```
Not related to this PR, but those constant objects `objectSizePutMethods` & `objectSizeGetMethods` should be declared only once outside the function to avoid reallocation of constant objects for each requests.

leif-scality replied:
moved variable to global scope

BourgoisMickael commented on lib/utilities/serverAccessLogger.js:
```diff
@@ -182,74 +182,78 @@ function getRemoteIPFromRequest(request) {
+    // https://github.com/awslabs/glue-extensions-for-iceberg/blob/52bdb2908216a85859fd76a45981d0326d016a2f/spark/src/main/scala/software/amazon/glue/s3a/audit/S3LogVerbs.java
+    // https://github.com/open-io/swift/blob/ff518e9907f74b5a2565973a260f36386b5d5cbf/etc/s3-default.cfg.in#L78
+    // https://stackoverflow.com/questions/42707878/amazon-s3-logs-operation-definition
     const methodToResType = Object.freeze({
```
This constant object should be declared only once outside the function to avoid reallocation for each requests. (it's a quit big object it can have a small impact)

leif-scality replied:
moved variable to global scope

BourgoisMickael reviewed (commented):
getBytesSent could be simplfied to 

```js

function getBytesSent(res, bytesSent) {
    if (bytesSent !== undefined && bytesSent !== null) {
        return bytesSent;
    }
    return res?.getHeader('Content-Length') || null;
}
```

leif-scality commented:
ping

leif-scality commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/9.2`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.8`
* `development/9.0`
* `development/9.1`



Please check the status of the associated issue CLDSRV-780.

Goodbye leif-scality.




*The following options are set:* **approve**
